math/src/elliptic_curve/short_weierstrass/curves/bls12_381/twist.rs

```diff
@@ -46,7 +47,72 @@ impl IsShortWeierstrass for BLS12381TwistCurve {
+            <Self as IsEllipticCurve>::BaseField::mul(&y_sq, z.value())
+        };
+
+        let rhs = {
+            let x_sq = <Self as IsEllipticCurve>::BaseField::square(x.value());
+            let x_cubed = <Self as IsEllipticCurve>::BaseField::mul(&x_sq, x.value());
+
+            let z_sq = <Self as IsEllipticCurve>::BaseField::square(z.value());
+            let z_cubed = <Self as IsEllipticCurve>::BaseField::mul(&z_sq, z.value());
+
+            let z_cubed_with_b =
+                <Self as IsEllipticCurve>::BaseField::mul(&z_cubed, Self::b().value());
+
+            <Self as IsEllipticCurve>::BaseField::add(&x_cubed, &z_cubed_with_b)
+        };
+
+        lhs == rhs
+    }
+}
 impl ShortWeierstrassProjectivePoint<BLS12381TwistCurve> {
+    /// Returns the "Untwist-Frobenius-Twist" endomorphism
+    pub fn psi(&self) -> Self {
+        // Coefficients borrowed from https://github.com/Consensys/gnark-crypto/blob/55489ac07ca1a88bf0b830a29625fcb0d8879a48/ecc/bls12-381/bls12-381.go#L132C1-L135C138
+        let psi_coeff_x = FieldElement::<<BLS12381TwistCurve as IsEllipticCurve>::BaseField>::new([
+            FieldElement::zero(),
+            FieldElement::new(U384::from_hex_unchecked("1A0111EA397FE699EC02408663D4DE85AA0D857D89759AD4897D29650FB85F9B409427EB4F49FFFD8BFD00000000AAAD"))
+        ]);
+
+        let psi_coeff_y = FieldElement::<<BLS12381TwistCurve as IsEllipticCurve>::BaseField>::new([
+            FieldElement::new(U384::from_hex_unchecked("135203E60180A68EE2E9C448D77A2CD91C3DEDD930B1CF60EF396489F61EB45E304466CF3E67FA0AF1EE7B04121BDEA2")),
+            FieldElement::new(U384::from_hex_unchecked("6AF0E0437FF400B6831E36D6BD17FFE48395DABC2D3435E77F76E17009241C5EE67992F72EC05F4C81084FBEDE3CC09"))
+        ]);
+
+        let frob_map_x = Degree2ExtensionField::frobenius_map(self.x().value());
+        let frob_map_x_with_psi_coeff_x =
+            Degree2ExtensionField::mul(&frob_map_x, psi_coeff_x.value());
+
+        let frob_map_y = Degree2ExtensionField::frobenius_map(self.y().value());
+        let frob_map_y_with_psi_coeff_y =
+            Degree2ExtensionField::mul(&frob_map_y, psi_coeff_y.value());
+
+        let frob_map_z = Degree2ExtensionField::frobenius_map(self.z().value());
+
+        <BLS12381TwistCurve as IsEllipticCurve>::PointRepresentation::new([
+            FieldElement::new(frob_map_x_with_psi_coeff_x),
+            FieldElement::new(frob_map_y_with_psi_coeff_y),
+            FieldElement::new(frob_map_z),
+        ])
+    }
+
+    /// Returns true if the point is in the prime subgroup `G_2` of order `r`.
+    /// Makes use of endomorphism for efficient point multiplication.
+    pub fn is_in_subgroup(&self) -> bool {
+        let seed_times_p = self.operate_with_self(SEED);
+        let psi_plus_seed_times_p = self.psi().operate_with(&seed_times_p);
+
+        Degree2ExtensionField::is_zero(psi_plus_seed_times_p.z().value())
+            && BLS12381TwistCurve::is_on_curve(&psi_plus_seed_times_p)
```

Comment:
I saw that's how its implemented in [Gnark](https://github.com/Consensys/gnark-crypto/blob/ace5318b726c7185249bc9c92e1cadae701c1654/ecc/bls12-381/g2.go#L405). I should have listened to it that both checks for same thing.

Reply:
Actually it doesn't really have to do with `z` equal to zero implying being equal to the point at infinity (since that already assumes that the point is on the curve). We should be able to rely on the fact that points of type `ShortWeierstrassProjectivePoint<BLS12381TwistCurve>` are on the curve. So a method such as `is_on_curve` should be redundant for instances of that type.

Reply:
Yeah, it is because a point which may not be on curve will have z-coordinate set at 0

Reply:
Maybe, `ShortWeierstrassProjectivePoint` can be made public at the crate level and add a builder fn that checks whether the point is on curve. Will that work?

Reply:
Yes. Prolly we only need the two builder functions we already have: [`create_point_from_affine`](https://github.com/lambdaclass/lambdaworks/blob/314fafcf44e1764dafa748f82d6df045f309cbc6/math/src/elliptic_curve/traits.rs#L26-L32) and [`deserialize` ](https://github.com/lambdaclass/lambdaworks/blob/314fafcf44e1764dafa748f82d6df045f309cbc6/math/src/elliptic_curve/short_weierstrass/point.rs#L213-L250). Both are checking the defining equation already.
But the attribute of the projective point is set to [public](https://github.com/lambdaclass/lambdaworks/blob/314fafcf44e1764dafa748f82d6df045f309cbc6/math/src/elliptic_curve/short_weierstrass/point.rs#L15C64-L15C86). So probably one could instantiate a mutable point from outside this crate and later change its coordinates. That would brake the "is on curve" invariant. That's a problem.

You can assume in this PR that any instance of type `ShortWeierstrassProjectivePoint<E>` is a point that's on the curve `E`. Let's discuss the other issue somewhere else. Feel free to create an issue for it (if you can provide an example and a candidate solution, even better!).